### PR TITLE
Add audio debrief recording flow

### PR DESCRIPTION
**Pull request by mathieuhendey**

### CODE REVIEW

File: config.xml [hunk 1]
```diff
@@ -96,4 +95,10 @@
     </plugin>
     <plugin name="cordova-plugin-device-authentication" spec="~0.0.1" />
     <plugin name="cordova-plugin-globalization" spec="^1.11.0" />
+    <plugin name="cordova-plugin-file" spec="^6.0.1" />
+    <plugin name="cordova-plugin-media" spec="^5.0.2" />
+    <allow-navigation href="http://172.27.237.231:8100" />
+    <allow-navigation href="http://192.168.14.205:8100" />
```
Comment (zielinskikamil) [inline config.xml]: these ips match your local network, these might not to work with on other machines/networks

File: ionic.config.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "App",
-  "app_id": "",
+  "app_id": "b07b5981",
```
Comment (zielinskikamil) [inline ionic.config.json]: we shouldn't set app id

File: package.json [hunk 3]
```diff
@@ -32,27 +32,32 @@
...
     "angular2-signaturepad": "^2.8.0",
+    "cordova-browser": "5.0.3",
     "cordova-ios": "4.5.4",
-    "cordova-plugin-advanced-http": "^1.11.0",
-    "cordova-plugin-device": "^2.0.1",
+    "cordova-plugin-advanced-http": "^1.11.1",
+    "cordova-plugin-device": "^2.0.2",
     "cordova-plugin-device-authentication":
-      "https://github.com/dvsa/cordova-plugin-device-authentication.git",
+      "git+https://github.com/dvsa/cordova-plugin-device-authentication.git",
```
Comment (zielinskikamil) [inline package.json]: git+ prefix make that app wont build on ionic pro. @danielchildsbjss  Am I right?

File: package.json [hunk 4]
```diff
@@ -102,8 +107,10 @@
       "cordova-plugin-screen-orientation": {},
       "cordova-plugin-insomnia": {},
       "cordova-plugin-device-authentication": {},
-      "cordova-plugin-globalization": {}
+      "cordova-plugin-globalization": {},
+      "cordova-plugin-file": {},
+      "cordova-plugin-media": {}
     },
-    "platforms": ["ios"]
+    "platforms": ["ios", "browser"]
```
Comment (zielinskikamil) [inline package.json]: hym, do we support browser?

File: src/app/app.component.ts [hunk 5]
```diff
@@ -1,14 +1,15 @@
...
 import { StatusBar } from '@ionic-native/status-bar';
 import { SplashScreen } from '@ionic-native/splash-screen';
-import { LoginPage } from '../pages/login/login';
 import { Content } from 'ionic-angular/navigation/nav-interfaces';
 import { DEFAULT_LANG, SYS_OPTIONS, AVAILABLE_LANG } from './constants';
 import { TranslateService } from 'ng2-translate';
 import { ScreenOrientation } from '@ionic-native/screen-orientation';
 import { Insomnia } from '@ionic-native/insomnia';
 import { Globalization } from '@ionic-native/globalization';
+// import { TestResultPage } from '../pages/test-result/test-result';
```
Comment (zielinskikamil) [inline src/app/app.component.ts]: ups some comment

File: src/components/record-debrief-button/record-debrief-button.scss [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+          -webkit-box-shadow: 0 0 0 0 rgba(204, 44, 44, 0.4);
+        }
+        70% {
+          -webkit-box-shadow: 0 0 0 10px rgba(204, 44, 44, 0);
+        }
+        100% {
+          -webkit-box-shadow: 0 0 0 0 rgba(204, 44, 44, 0);
+        }
+      }
+      @keyframes pulse {
```
Comment (zielinskikamil) [inline src/components/record-debrief-button/record-debrief-button.scss]: like this animations +1

File: src/components/record-debrief-button/record-debrief-button.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+@Component({
+  selector: 'record-debrief-button',
+  templateUrl: 'record-debrief-button.html'
+})
+export class RecordDebriefButtonComponent {
+  @Input() isRecording: boolean;
+  @Input() isPausedRecording: boolean;
+  @Input() activeFile: any;
+  constructor() {
+    // if(!this.isRecording){
```
Comment (zielinskikamil) [inline src/components/record-debrief-button/record-debrief-button.ts]: whole constructor can be removed

File: src/pages/test-result/test-result.ts [hunk 8]
```diff
@@ -19,15 +21,45 @@ export class TestResultPage {
...
+    private faultStore: FaultStoreProvider,
+    private file: File,
+    private toastCtrl: ToastController
   ) {
     this.faultStore
       .getFaultTotals()
       .subscribe((faultSummaries) => (this.faultSummaries = faultSummaries));
     this.testResult = this.faultStore.getTestResult();
+    this.debriefConsent = this.faultStore.getDebriefConsentStatus();
+    // this.debriefConsent = true;
```
Comment (zielinskikamil) [inline src/pages/test-result/test-result.ts]: another comment

File: src/pages/test-result/test-result.ts [hunk 9]
```diff
@@ -39,4 +71,154 @@ export class TestResultPage {
...
   }
+
+  toggleStopOrDestroy = () => {
+    if (this.stopOrDestroyRecording === 'stop') {
+      this.stopRecordingAudio();
+    }
+    if (this.stopOrDestroyRecording === 'destroy') {
+      this.deleteAudio();
+    }
+    this.stopOrDestroyRecording = this.stopOrDestroyRecording === 'stop' ? 'destroy' : 'stop';
```
Comment (zielinskikamil) [inline src/pages/test-result/test-result.ts]: this looks a bit weird ^^

File: src/pages/test-result/test-result.ts [hunk 10]
```diff
@@ -39,4 +70,154 @@ export class TestResultPage {
...
+          this.playOrPauseRecording = 'pause';
+        })
+        .catch(this.logError);
+    }
+  }
+
+  stopRecordingAudio() {
+    console.log('Stopping recording');
+
+    const self = this;
```
Comment (danielchildsbjss) [inline src/pages/test-result/test-result.ts]: We shouldn't need to do this anymore, when we use => for functions, the this scope makes sense

File: src/pages/test-result/test-result.ts [hunk 11]
```diff
@@ -39,4 +70,154 @@ export class TestResultPage {
...
+        this.pauseRecording();
+      } else {
+        this.pauseAudio();
+      }
+    }
+    this.playOrPauseRecording = this.playOrPauseRecording === 'play' ? 'pause' : 'play';
+  };
+
+  // audio debrief
+  recordAudio() {
```
Comment (danielchildsbjss) [inline src/pages/test-result/test-result.ts]: All this record logic should be in a service really, not in a page

File: src/pages/test-result/test-result.ts [hunk 12]
```diff
@@ -39,4 +70,154 @@ export class TestResultPage {
...
+      const dur = this.audio.getDuration();
+      if (dur > 0) {
+        clearInterval(timerDur);
+        this.fileLength = dur;
+      }
+    }, 100);
+  }
+
+  playAudio() {
+    const self = this;
```
Comment (danielchildsbjss) [inline src/pages/test-result/test-result.ts]: Here too, don't do this, just use this like normal

File: config.xml [hunk 13]
```diff
@@ -96,4 +95,13 @@
...
     <plugin name="cordova-plugin-device-authentication" spec="~0.0.1" />
     <plugin name="cordova-plugin-globalization" spec="^1.11.0" />
+    <plugin name="cordova-plugin-file" spec="^6.0.1" />
+    <plugin name="cordova-plugin-media" spec="^5.0.2" />
+    <edit-config file="*-Info.plist" mode="merge" target="NSMicrophoneUsageDescription">
+        <string>Need microphone access to record audio</string>
+    </edit-config>
+    <engine name="ios" spec="~4.5.4" />
+    <engine name="browser" spec="~5.0.3" />
+    <allow-navigation href="http://192.168.14.202:8100" />
```
Comment (danielchildsbjss) [inline config.xml]: Can we remove ip addresses

File: src/pages/declaration-consent/declaration-consent.ts [hunk 14]
```diff
@@ -78,11 +80,13 @@ export class DeclarationConsentPage {
       .runAuthentication('Please authenticate yourself to proceed')
       .then((isAuthenticated: boolean) => {
         if (isAuthenticated) {
+          this.faultStore.setDebriefConsentStatus(this.checkDebrief);
```
Comment (danielchildsbjss) [inline src/pages/declaration-consent/declaration-consent.ts]: These aren't really faults, should we store them here?

File: src/pages/test-result/test-result.ts [hunk 15]
```diff
@@ -19,15 +20,45 @@ export class TestResultPage {
     [key: string]: IFaultSummary;
   };
 
+  // audio related
```
Comment (danielchildsbjss) [inline src/pages/test-result/test-result.ts]: All this audio stuff could go in your provider if you wanted

Comment (fredski02) [reply src/pages/test-result/test-result.ts]: Provider has been added now :)

File: src/pages/declaration-consent/declaration-consent.ts (hunk 14 above)
Comment (fredski02) [reply src/pages/declaration-consent/declaration-consent.ts]: sausages

**Pull request merged**